### ISSUE TITLE
Random initial conditions with other distributions

### ISSUE THREAD

**Issue by bwspenc** [idaholab/moose#9710]
### Description of the enhancement or error report
We currently have an IC called RandomIC that permits initializing a variable using a uniform random distribution. Expand this to a wider set of distributions.

### Rationale for the enhancement or information for reproducing the error
I'm specifically interested in an element volume-weighted Weibull distribution such as the one documented here:
http://doi.wiley.com/10.1002/nme.4699
to represent local variations in the material's strength for use in failure models. I imagine that other distributions might be of interest as well.

### Identified impact
Either add a set of other IC objects or hook the RandomIC up with a distribution object.

**Comment by bwspenc** [comment]
@aeslaughter I'm not sure whether we should create a family of IC objects for various distributions, or hook the IC object up with a distribution object. I'd like to discuss this with you.  I think it would be nice to use a distribution object, except that in this case, the sampled value is dependent on the element size, so maybe that doesn't make a whole lot of sense.

I think the variant of the Weibull distribution that I want here might be specific enough to mechanics that it should go in tensor_mechanics rather than the framework.

Also, the current `RandomIC` object might need its name changed to something more specific like `UniformRandomIC` if we add more of these.

I have a branch that implements an initial version of this.  I'll tag it with this issue number.

**Comment by aeslaughter** [comment]
> I think it would be nice to use a distribution object

I agree, this should use a distribution object. That way you are not restricted to Weibull, you could use any distribution.

**Comment by permcody** [comment]
Another thing that would be nice to fix is parallel reproducibility with the RandomIC object. It's currently the only original Random classes that doesn't work properly. The reason is that the execution path over the IC points is complicated. I tried to fix it once and failed, but I think it's possible. Not necessary for this ticket per se, but I just wanted to get people people thinking about it.

**Comment by GiudGiud** [comment]
Closing as addressed by https://github.com/idaholab/moose/pull/12345
